src/style/properties.js

```diff
@@ -0,0 +1,645 @@
+        this.begin = now + transition.delay || 0;
+        this.end = this.begin + transition.duration || 0;
+        if (transition.delay || transition.duration) {
+            this.prior = prior;
+        }
+    }
+
+    possiblyEvaluate(parameters: EvaluationParameters): R {
+        const now = parameters.now || 0;
+        const finalValue = this.value.possiblyEvaluate(parameters);
+        const prior = this.prior;
+        if (!prior) {
+            // No prior value.
+            return finalValue;
+        } else if (now > this.end) {
+            // Transition from prior value is now complete.
+            this.prior = null;
+            return finalValue;
+        } else if (this.value.isDataDriven()) {
+            // Transitions to data-driven properties are not supported.
+            // We snap immediately to the data-driven value so that, when we perform layout,
+            // we see the data-driven function and can use it to populate vertex buffers.
+            this.prior = null;
+            return finalValue;
+        } else if (now < this.begin) {
+            // Transition hasn't started yet.
+            return prior.possiblyEvaluate(parameters);
+        } else {
+            // Interpolate between recursively-calculated prior value and final.
+            const t = (now - this.begin) / (this.end - this.begin);
+            return this.property.interpolate(prior.possiblyEvaluate(parameters), finalValue, easeCubicInOut(t));
+        }
+    }
+}
+
+/**
+ * A helper type: given an object type `Properties` whose values are each of type `Property<T, R>`, it calculates
+ * an object type with the same keys and values of type `TransitioningPropertyValue<T, R>`.
+ *
+ * @private
+ */
+type TransitioningPropertyValues<Properties: Object>
+    = $Exact<$ObjMap<Properties, <T, R>(p: Property<T, R>) => TransitioningPropertyValue<T, R>>>
+
+/**
+ * `Transitioning` stores a map of all (property name, `TransitioningPropertyValue`) pairs for paint properties of a
+ * given layer type. It can calculate the possibly-evaluated values for all of them at once, producing a
+ * `PossiblyEvaluated` instance for the same set of properties.
+ *
+ * @private
+ */
+class Transitioning<Properties: Object> {
+    _values: TransitioningPropertyValues<Properties>;
+
+    constructor(values: TransitioningPropertyValues<Properties>) {
+        this._values = values;
+    }
+
+    possiblyEvaluate(parameters: EvaluationParameters): PossiblyEvaluated<Properties> {
+        const result: any = {};
+        for (const property in this._values) {
+            result[property] = this._values[property].possiblyEvaluate(parameters);
+        }
+        return new PossiblyEvaluated(result); // eslint-disable-line no-use-before-define
+    }
+
+    hasTransition() {
+        for (const property in this._values) {
+            if (this._values[property].prior) {
+                return true;
+            }
+        }
+        return false;
+    }
```

Comment:
Along the lines of the point above re: the temporal state in `TransitioningPropertyValue`, would it make sense for `hasTransition()` to accept a `now` parameter (here and elsewhere)?